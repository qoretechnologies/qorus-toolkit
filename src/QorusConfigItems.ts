export interface IConfigData {
  configItems: any[];
  config: Record<string, any>;
}

class QorusConfigItems {
  configItems;

  constructor(configData: IConfigData) {
    this.configItems = this.parseConfigItems(configData);
  }

  private parseConfigItems(configData: IConfigData) {
    const { config, configItems } = configData;
    const jobData =
      config &&
      Object.keys(config)?.map((configKey) => {
        const configItem = configItems?.find((configItemData) => configItemData.name === configKey);

        const obj = {
          name: configKey,
          type: config[configKey].type,
          canBeNull: config[configKey].type?.includes(`*`),
          description: config[configKey].desc,
          strictyLocal: config[configKey].stricty_local,
          configGroup: config[configKey].config_group,
          sensitive: config[configKey].sensitive,
          prefix: config[configKey].prefix,
          value: config[configKey].value,
          level: config[configKey].level,
          isSet: config[configKey].is_set,
          isDefaultValueTemplatedString: configItem?.is_default_value_templated_string,
          isValueTemplatedString: configItem?.is_value_templated_string,
          allowedValues: configItem?.allowed_values ?? null,
        };
        return obj;
      });
    return jobData;
  }

  getAll() {
    return this.configItems;
  }

  get(configItemName: string) {
    return this.configItems.find((item) => item.name === configItemName);
  }
}
export default QorusConfigItems;
