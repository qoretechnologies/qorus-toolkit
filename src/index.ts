export * as Authenticator from './QorusAuthenticator';
export * as HttpRequest from './QorusRequest';
export * as DataProvider from './QorusDataProvider';
export * as QorusOptions from './QorusOptions';
export * as Validator from './QorusValidator';
export * as Jobs from './QorusJobs';
export * as Workflows from './QorusWorkflows';
export * as QorusInterface from './QorusInterface';

export { default as QorusAuthenticator } from './QorusAuthenticator';
export { default as QorusRequest } from './QorusRequest';
export { default as QorusDataProvider } from './QorusDataProvider';
export { default as QorusValidator } from './QorusValidator';
export { default as QorusJobs } from './QorusJobs';
export { default as QorusWorkflows } from './QorusWorkflows';

export { TVersion, IApiPaths, IApiPathsAuthenticator, IApiPathsDataProvider, IApiPathsJobs } from './utils/apiPaths';
export { TContext } from './QorusDataProvider';
export { IDefaultHeaders } from './QorusRequest';

/**
 * A record of objects with string key and string value
 */
export type TObjectWithStringKey = Record<string, string>;

/**
 * A record of objects with string key and any kind of value
 */
export type TObjectWithAnyValue = Record<string, any>;
