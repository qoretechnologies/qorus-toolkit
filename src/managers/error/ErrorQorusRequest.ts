import BaseError from './Error';
import { errorCodes } from './errorCodes';

class ErrorQorusRequest extends BaseError {
  constructor(error: IErrorQorusRequestParams) {
    if (typeof error.desc === 'undefined') {
      super(`${JSON.stringify(error)}`, true, errorCodes.INTERNAL.name, undefined);
    } else {
      super(error.desc, true, error.err, error.status);
    }
  }
}

export interface IErrorQorusRequestParams {
  desc: string;
  err: string;
  status: number;
}

export default ErrorQorusRequest;
