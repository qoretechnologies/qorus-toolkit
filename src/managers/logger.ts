import { Roarr as logger } from 'roarr';
import { detect } from 'detect-browser';
import { addUncaughtListener, startUncaughtListener } from '../utils/catchUncaughtErrors';
import { ROARR } from 'roarr';

const browser = detect();

if (browser?.type !== 'node') {
  ROARR.write = (message) => {
    logger.info(JSON.stringify(message));
  };
}

startUncaughtListener();
addUncaughtListener((error) => {
  logger.error(error);
});

export default logger;
