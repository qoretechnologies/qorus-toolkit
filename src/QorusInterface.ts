import ErrorQorusRequest, { IErrorQorusRequestParams } from './managers/error/ErrorQorusRequest';
import QorusRequest, { QorusRequestResponse } from './QorusRequest';
import { getRequestPath } from './utils';

/**
 * QorusInterface is a master class used by different Qorus Interfaces
 * - Allows to call available actions on an Qorus interface
 * @returns QorusInterface class object
 * @Category QorusInterface
 */
export class QorusInterface {
  /**
   * Browse path for an interface
   */
  browsePath = '';

  constructor(browsePath: string) {
    this.browsePath = browsePath;
  }

  /**
   * A getter to get all jobs or a particular job from Qorus server job api
   * @param jobId Optional job id of a QorusJob
   * @returns List of jobs if job id is not provided, job data if the job id is provided, undefined if the job is not find
   */
  protected async _get<T>(param?: TQorusInterfaceIdentifier): Promise<T[] | T> {
    const requestPath = getRequestPath([this.browsePath ?? '', param?.toString() ?? '']);
    const response = await QorusRequest.get({ path: requestPath });

    const responseObj = response as QorusRequestResponse;
    const error = response as IErrorQorusRequestParams;

    if (error.status) {
      throw new ErrorQorusRequest(error);
    }
    const responseData: T = responseObj?.data as T;

    return responseData;
  }

  /**
   * Enable an interface on a Qorus server
   * @param param Id or name of the interface to be enabled
   * @returns EnableDisableResponse with jobId, name, version of the job
   */
  async enable(
    qorusInterfaceIdentifier: TQorusInterfaceIdentifier,
  ): Promise<IQorusEnableDisableTransformed | undefined> {
    const response = await this.makeActionCall('enable', qorusInterfaceIdentifier, true);
    return response as IQorusEnableDisableTransformed;
  }

  /**
   * Enable an interface on a Qorus server
   * @param param Id or name of the interface to be enabled
   * @returns EnableDisableResponse with jobId, name, version of the job
   */
  async disable(
    qorusInterfaceIdentifier: TQorusInterfaceIdentifier,
  ): Promise<IQorusEnableDisableTransformed | undefined> {
    const response = await this.makeActionCall('enable', qorusInterfaceIdentifier, true);
    return response as IQorusEnableDisableTransformed;
  }

  /**
   * Kill an interface processes on a Qorus server
   * @param param Id or name of the interface to be enabled
   * @returns EnableDisableResponse with jobId, name, version of the job
   */
  async kill(qorusInterfaceIdentifier: TQorusInterfaceIdentifier) {
    const response = await this.makeActionCall('kill', qorusInterfaceIdentifier, true, 'post');
    return response as IQorusInterfaceKill;
  }

  /**
   * A helper method to create a action call to an Qorus server interface
   * @param param Id or name of the interface to be enabled
   * @returns Response from the action call, if the response has variable id names ex:'jobid', it
   * will return a transformed result with the variable id name changed to just "id"
   */
  protected async makeActionCall(
    actionName: string,
    qorusInterfaceIdentifier: TQorusInterfaceIdentifier,
    transformResponse = false,
    requestType: 'post' | 'put' = 'put',
  ): Promise<
    | IQorusInterfaceEnableDisableResponse
    | IQorusEnableDisableTransformed
    | IQorusInterfaceKill
    | Record<string, string | number | boolean>
    | undefined
  > {
    if (requestType) {
    }
    const requestPath: string = getRequestPath([this.browsePath, qorusInterfaceIdentifier?.toString()]);
    const requestParams = {
      action: actionName,
    };
    const response = await QorusRequest[requestType]({ path: requestPath, params: requestParams });
    const error = response as IErrorQorusRequestParams;
    if (error.status) {
      throw new ErrorQorusRequest(error);
    }
    const responseAsData = response as QorusRequestResponse;
    const responseObj = responseAsData.data as IQorusInterfaceEnableDisableResponse;
    return transformResponse ? this.transformInterface(responseObj) : responseObj;
  }

  protected transformInterface(
    interfaceEnableDisableResponse: IQorusInterfaceEnableDisableResponse,
  ): IQorusEnableDisableTransformed | undefined {
    if (
      interfaceEnableDisableResponse.hasOwnProperty('jobid') ||
      interfaceEnableDisableResponse.hasOwnProperty('workflowid')
    ) {
      const id = interfaceEnableDisableResponse.jobid ?? interfaceEnableDisableResponse.workflowid ?? '';
      const transformed: IQorusEnableDisableTransformed = {
        id,
        name: interfaceEnableDisableResponse.name,
        version: interfaceEnableDisableResponse.version,
        info: interfaceEnableDisableResponse.info,
      };
      return transformed;
    } else return undefined;
  }
}

/**
 *  Id or name of an Qorus interface
 */
export type TQorusInterfaceIdentifier = number | TInterfaceIdentifierName;

/**
 * Name of an Qorus interface
 */
export type TInterfaceIdentifierName = string;

export interface IQorusInterfaceKill {
  status: string;
  code: number;
}

/**
 * Transformed object to provide a general hash for the enable and disable action
 * for an Qorus server interface
 */
export interface IQorusEnableDisableTransformed {
  /**
   * Id of the interface
   */
  id: string;

  /**
   * Name of the interface
   */
  name: string;

  /**
   * Version of the interface
   */
  version: string;

  /**
   * Info about the interface after action call
   */
  info: string;
}

export interface IQorusInterfaceEnableDisableResponse {
  /**
   * Id of the job that has been enabled
   */
  jobid?: string;

  /**
   * Id of the workflow
   */
  workflowid?: string;

  /**
   * Name of the job that has been enabled
   */
  name: string;

  /**
   * Version of the job that has been enabled
   */
  version: string;

  /**
   * Info of the job that has been enabled
   */
  info: string;
}
