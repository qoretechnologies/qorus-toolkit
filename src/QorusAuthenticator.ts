import ErrorInternal from './managers/error/ErrorInternal';
import { getKeyValLocal, setKeyValLocal } from './managers/LocalStorage';
import logger from './managers/logger';
import QorusRequest, { QorusRequestResponse } from './QorusRequest';
import QorusValidator from './QorusValidator';
import { isValidString, isValidStringArray } from './utils';
import {
  apiPathsInitial,
  createApiPaths,
  IApiPaths,
  IApiPathsAuthenticator,
  IWithEndpointVersion,
  TVersion,
} from './utils/apiPaths';

/**
 * Endpoint id for a Qorus server endpoint
 */
export type TQorusEndpointId = string;

/**
 * Authentication token for a Qorus server endpoint
 */
export type TQorusAuthToken = string;

/**
 * Url for a Qorus server endpoint
 */
export type TQorusEndpointURL = string;

export interface IWithQorusEndpointId {
  /**
   * Id for a endpoint provided by the user, unique for every endpoint
   */
  endpointId: TQorusEndpointId;
}

export interface IWithQorusAuthToken {
  /**
   * Authentication token for the user provided endpoint
   */
  authToken?: TQorusAuthToken;
}

export interface IWithQorusURL {
  /**
   * URL to Qorus server for the provided endpoint
   */
  url: TQorusEndpointURL;
}

export interface ILoginParams {
  /**
   * Username for the authentication to Qorus server
   */
  user?: string;
  /**
   * Password for the authentication to Qorus server
   */
  pass?: string;
}

export interface IAddEndpoint extends IWithQorusURL, IWithEndpointVersion, IWithQorusEndpointId {}

export interface IEndpoint extends IWithQorusURL, IWithEndpointVersion, IWithQorusAuthToken, IWithQorusEndpointId {}

/**
 * Authentication token for a Qorus Endpoint
 */
export type TToken = string;

/**
 * QorusAuthenticator class provides methods to authenticate a user for a Qorus server endpoint
 * - Add Multiple endpoint for different versions of Qorus apis
 * - Supports no-auth for the Qorus server api
 * - Uses locally stored token or user and password for authentication
 * @returns QorusAuthenticator class object
 * @Category QorusAuthenticator
 */
export class QorusAuthenticator {
  /**
   * Array of user defined endpoints
   */
  endpoints: IEndpoint[] = [];

  /** Object of Api paths for the selected endpoint */
  allApiPaths: IApiPaths = apiPathsInitial;

  /** Api paths for the QorusAuthenticator */
  apiPathsAuthenticator: IApiPathsAuthenticator = apiPathsInitial.authenticator;

  /** Selected endpoint from the endpoints array */
  selectedEndpoint: IEndpoint | undefined;

  /** No auth identifier to identify if the no-auth is enabled for the user */
  noauth = false;

  /**
   * A getter to get the endpoint if it exist in the Endpoints array
   * @param id ID of the endpoint ex: "rippy"
   * @returns Endpoint object if the endpoint with the provided id exist in the endpoints array, undefined otherwise.
   */
  getEndpointById(endpointId: string): IEndpoint | undefined {
    return this.endpoints.find((endpoint) => endpoint.endpointId === endpointId);
  }

  /**
   * Logs out the current user from the selected endpoint
   * @returns True if the operation is successful, False otherwise
   */
  async logout(): Promise<boolean> {
    if (!this.selectedEndpoint || !isValidString(this.selectedEndpoint.authToken)) {
      return true;
    }

    try {
      await QorusRequest.post({ path: `${this.apiPathsAuthenticator.logout}` });
      return true;
    } catch (error: any) {
      throw new ErrorInternal(`Unable to logout user from the endpoint ${this.selectedEndpoint}`);
    } finally {
      this.selectedEndpoint.authToken = undefined;
      this.allApiPaths = apiPathsInitial;
      this.apiPathsAuthenticator = apiPathsInitial.authenticator;
      this.noauth = false;
    }
  }

  /**
   * Select an endpoint from the available Endpoints array
   * @param id Id of the endpoint
   * @returns Endpoint if the operation is successful, undefined otherwise.
   */
  async selectEndpoint(id: string): Promise<IEndpoint | undefined> {
    if (!isValidString(id)) {
      throw new ErrorInternal('Id is not valid, please provide a valid id or initialize a new endpoint.');
    }

    const endpoint = this.getEndpointById(id);
    if (!endpoint || !isValidString(endpoint.url)) {
      throw new ErrorInternal('Selected endpoint is not valid, please create a new endpoint.');
    }

    if (this.selectedEndpoint?.authToken) {
      await this.logout();
    }

    this.selectedEndpoint = endpoint;
    this.allApiPaths = createApiPaths({ version: endpoint.version });
    this.apiPathsAuthenticator = this.allApiPaths.authenticator;

    return endpoint;
  }

  /**
   * Checks if the Qorus endpoint supports no-auth
   * @param endpoint Endpoint config to add the data
   * @returns True if the no-auth is enabled for the user, False otherwise
   */
  async checkNoAuth(endpoint?: IEndpoint): Promise<boolean> {
    const actualEndpoint = endpoint ?? this.selectedEndpoint;

    /* Throwing an error if the actualEndpoint is not defined. */
    if (!actualEndpoint) {
      throw new ErrorInternal('No endpoint selected, please select an endpoint.');
    }

    /* Trying to connect to the endpoint and check if the endpoint is using noauth. */
    try {
      const resp = await QorusRequest.get<{ data: { noauth: boolean } }>(
        { path: `${this.apiPathsAuthenticator.validateNoAuth}` },
        endpoint,
      );
      if (!resp?.data) {
        throw new ErrorInternal(
          `Cannot verify no-auth please check your url "${endpoint?.url ?? this.selectedEndpoint?.url}" and try again`,
        );
      }

      if (!resp.data.noauth) {
        this.noauth = false;
        console.log('No auth disabled, authentication is required with username and password');
        return false;
      }

      if (resp.data.noauth) {
        this.noauth = resp.data.noauth;
        console.log('No auth enabled, authentication not required');
        return true;
      }

      this.noauth = false;

      return false;
    } catch (error: any) {
      throw new ErrorInternal(
        `Unable to connect to ${actualEndpoint.url}, please check the url / connection and try again. ${
          error ? JSON.stringify(error) : ''
        }`,
      );
    }
  }

  /**
   * A getter to get selected Endpoint
   * @returns Selected Endpoint if the endpoint exists, undefined otherwise
   */
  getSelectedEndpoint(): IEndpoint | undefined {
    return this.selectedEndpoint;
  }

  /**
   * Fixes the endpoint data
   * @param data AddEndpoint, data to be fixed
   * @returns Fixed Endpoint config
   */
  #fixEndpointData(data: IAddEndpoint): IAddEndpoint & ILoginParams {
    const newData = { ...data };

    if (!newData.version) {
      newData.version = 'latest';
    }

    return newData;
  }

  /**
   * Checks the validity of the selected endpoint
   * @param data Endpoint data to be checked
   * @param withCredentials boolean to check if the endpoint has credentials
   * @returns True if the Endpoint data is valid, False otherwise
   */
  validateEndpointData(data: IAddEndpoint & ILoginParams, withCredentials?: boolean): boolean {
    let valid = true;
    const fixedData = this.#fixEndpointData(data);

    if (
      !(
        QorusValidator.validate('string', fixedData.endpointId) &&
        QorusValidator.validate('string', fixedData.url) &&
        QorusValidator.validate('version', fixedData.version)
      )
    ) {
      valid = false;
    }

    if (
      withCredentials &&
      !(QorusValidator.validate('string', fixedData.user) && QorusValidator.validate('string', fixedData.pass))
    ) {
      valid = false;
    }

    return valid;
  }

  /**
   * Validates the local stored authentication token for the Endpoint
   * @param endpointId Id of the endpoint
   * @returns Authentication token, if the authentication is successful, null otherwise
   */
  async validateLocalUserToken(endpointId: string): Promise<string | null> {
    const authToken = getKeyValLocal(`auth-token-${endpointId}`);

    if (!isValidString(authToken)) {
      return null;
    }

    try {
      const resp = await QorusRequest.get({
        path: `${this.apiPathsAuthenticator.validateToken}`,
        data: { token: authToken },
      });

      if (typeof resp === 'string') {
        return authToken;
      } else {
        return null;
      }
    } catch (error: any) {
      return null;
    }
  }

  /**
   * Authenticates the user to interact with the Qorus api.
   * If the username and password is not provided it tries to authenticate the user using the locally stored token from the selected Endpoint
   * @param loginParams LoginParams, user and pass is required to authenticate the user.
   * @returns Authentication token if the authentication is successful, undefined otherwise.
   */
  async login(loginParams?: ILoginParams): Promise<string | undefined> {
    await this.checkNoAuth();
    if (this.noauth) {
      logger.log('No-auth enabled authentication not required.');
      return undefined;
    }
    if (!this.selectedEndpoint || !isValidString(this.selectedEndpoint.url)) {
      throw new ErrorInternal('Endpoint must be initialized before authentication.');
    }
    if (!isValidStringArray([loginParams?.user, loginParams?.pass])) {
      throw new ErrorInternal('Username and password is required for authentication');
    }

    const user = loginParams?.user ?? undefined;
    const pass = loginParams?.pass ?? undefined;
    const { endpointId } = this.selectedEndpoint;
    const currentUserToken = await this.validateLocalUserToken(endpointId);

    if (currentUserToken && currentUserToken !== 'invalid' && this.selectedEndpoint) {
      this.selectedEndpoint.authToken = currentUserToken;
      return currentUserToken;
    }

    const resp = await QorusRequest.post({
      path: `${this.apiPathsAuthenticator.login}`,
      data: { user, pass },
    });
    const responseData = resp as QorusRequestResponse;
    if (typeof responseData.data === 'undefined') {
      throw new ErrorInternal(`${responseData}`);
    }
    const { token } = responseData?.data ?? null;
    if (!token) {
      throw new Error('There was an error authenticating user, token is invalid, please try again.');
    }

    this.selectedEndpoint.authToken = token;
    setKeyValLocal({ key: `auth-token-${endpointId}`, value: token });
    return token;
  }

  /**
   * Add a new Qorus Endpoint to interact with the qorus api
   * @param endpointConfig Endpoint configuration required to add a new endpoint
   * @returns Newly added endpoint
   */
  addEndpoint(endpointConfig: IAddEndpoint): IEndpoint {
    const { endpointId, url } = endpointConfig;
    const newEndpoint: IEndpoint = this.#fixEndpointData(endpointConfig);

    /* Checking if the id and url are valid strings. */
    if (!isValidStringArray([endpointId, url])) {
      throw new ErrorInternal('Id and url is required to initialize an endpoint');
    }

    /* Checking if the endpoint already exists. */
    if (this.getEndpointById(endpointId)) {
      throw new ErrorInternal(
        `Endpoint with the id "${endpointId}" already exists, please try again with a different id`,
      );
    }

    /* Adding a new endpoint to the endpoints array. */
    this.endpoints.push(newEndpoint);
    this.selectedEndpoint = newEndpoint;

    return this.selectedEndpoint;
  }

  /**
   * Allows the user to renew the selected endpoint authentication token
   * @param loginParams LoginParams optional username and password can be provided
   * @returns Token if the authentication is successful, undefined otherwise
   */
  async renewSelectedEndpointToken(loginParams: ILoginParams): Promise<string | undefined> {
    const { user, pass } = loginParams;
    if (!this.selectedEndpoint || !isValidString(this.selectedEndpoint.url)) {
      console.error(new ErrorInternal('Endpoint is not selected, please initialize an endpoint to renew token'));
      return undefined;
    }
    if (!isValidStringArray([user, pass])) {
      throw new ErrorInternal('Username and password is required to revalidate endpoint token.');
    }

    const token = await this.login({ user, pass });
    if (isValidString(token)) {
      this.selectedEndpoint.authToken = token;
      return token;
    }

    console.error(new ErrorInternal('Username and password are not valid, please try again'));
    return undefined;
  }

  /**
   * A getter to return the auth token of the selected Endpoint
   * @returns token if the the selected endpoint exists and the user is authenticated, otherwise returns undefined
   */
  getAuthToken(): string | undefined {
    return this.selectedEndpoint?.authToken;
  }

  /**
   * A getter to get the api Version of a Endpoint
   * @param endpointId Optional id parameter to get the version of a particular endpoint
   * @returns Version of the selected endpoint or version of the the endpoint found by id,
   * if the endpoint doesn't exists it returns undefined
   */
  getEndpointVersion(endpointId?: string): TVersion | undefined {
    if (isValidString(endpointId)) {
      return this.getEndpointById(endpointId!)?.version;
    } else {
      if (this.selectedEndpoint) {
        return this.selectedEndpoint.version;
      }
    }

    return undefined;
  }

  /**
   * Validates if the provided version is an accepted api version
   * @param version Version of the endpoint
   * @returns True if the version is valid, False otherwise
   */
  private validateVersion(version: number | string) {
    const versions = [1, 2, 3, 4, 5, 6, 'latest'];
    return versions.includes(version);
  }

  /**
   * A setter to set the Version of a Endpoint
   * @param version Version of the qorus api
   * @param endpointId Optional parameter to change the url of a particular endpoint from the endpoints array
   * @returns Version of the endpoint if the operation is successful, undefined otherwise
   *
   */
  async setEndpointVersion(version: TVersion, endpointId?: string): Promise<TVersion | undefined> {
    if (!endpointId) endpointId = this.selectedEndpoint?.endpointId;
    if (isValidString(endpointId) && this.validateVersion(version)) {
      const endpoint = this.getEndpointById(endpointId!);
      await this.logout();

      if (endpoint && isValidString(endpoint.url)) {
        this.endpoints[this.endpoints.indexOf(endpoint)].version = version;

        if (this.selectedEndpoint && this.selectedEndpoint.endpointId === endpoint.endpointId) {
          this.selectedEndpoint.version = version;
          this.allApiPaths = createApiPaths({ version });
          this.apiPathsAuthenticator = this.allApiPaths.authenticator;
        }

        logger.log('Changed endpoint version successfully.');
        return version;
      }

      throw new ErrorInternal('Endpoint does not exist, please initialize an endpoint and try again.');
    }

    throw new ErrorInternal('Id and a valid version is required to change the version of a endpoint.');
  }

  /**
   * A setter to set the url of the selected Endpoint
   * @param url Base url for the endpoint
   * @param id Optional id parameter to change the url of a particular endpoint
   * @returns Url of the endpoint if the operation is successful, undefined otherwise
   */
  async setEndpointUrl(url: string, endpointId?: string): Promise<string | undefined> {
    if (!isValidString(endpointId)) endpointId = this.selectedEndpoint?.endpointId;
    if (isValidStringArray([endpointId, url])) {
      const endpoint = this.getEndpointById(endpointId!);
      await this.logout();

      if (endpoint && isValidString(endpoint.url)) {
        this.endpoints[this.endpoints.indexOf(endpoint)].url = url;

        if (this.selectedEndpoint && this.selectedEndpoint.endpointId === endpoint.endpointId) {
          this.selectedEndpoint.url = url;
        }

        logger.log('Changed endpoint url successfully.');
        return url;
      }

      throw new ErrorInternal('Endpoint does not exist, please initialize an endpoint and try again.');
    }

    throw new ErrorInternal('Id and url is required to change the version of a endpoint.');
  }

  /**
   * A getter to return the api paths for the selected Endpoint
   * @returns ApiPaths for the selected endpoint if exists, otherwise returns default api paths
   */
  getApiPaths(): IApiPaths {
    return this.allApiPaths;
  }

  /**
   * A getter to get all the available Endpoints
   * @returns Endpoints array with all the available endpoints
   */
  getAllEndpoints(): IEndpoint[] {
    return this.endpoints;
  }
}

export default new QorusAuthenticator();
