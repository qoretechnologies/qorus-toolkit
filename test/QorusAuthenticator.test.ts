import dotenv from 'dotenv';
import { QorusAuthenticator } from '../src';
import ErrorInternal from '../src/managers/error/ErrorInternal';
import ErrorQorusRequest from '../src/managers/error/ErrorQorusRequest';
import logger from '../src/managers/logger';

dotenv.config();

const loggerMock = jest.spyOn(logger, 'log');

if (!(process.env.ENDPOINT && process.env.TESTUSER && process.env.TESTPASS)) {
  throw new Error('Missing required environment variables');
}

describe('QorusLogin Utility Class Tests', () => {
  jest.setTimeout(30000);
  it('Should initialize the endpoint and assign it to the selected endpoint', () => {
    QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT!, endpointId: 'rippy' });

    const endpoint = QorusAuthenticator.getSelectedEndpoint();
    expect(endpoint).toMatchSnapshot();
  });

  it('Should return user token after authentication (login)', async () => {
    let token = await QorusAuthenticator.login({ user: process.env.TESTUSER!, pass: process.env.TESTPASS! });

    expect(typeof token).toEqual('string');
  });

  it('Should return true after successfully loging out the user ', async () => {
    await QorusAuthenticator.login({ user: process.env.TESTUSER!, pass: process.env.TESTPASS! });
    const logoutResult: boolean = await QorusAuthenticator.logout();

    expect(logoutResult).toEqual(true);
  });

  it('Should return the enpoint from the endpoints array', () => {
    const endpoint = QorusAuthenticator.getEndpointById('rippy');

    expect(endpoint?.endpointId).toEqual('rippy');
  });

  it('Should return all the available endpoints ', () => {
    const endpoints = QorusAuthenticator.getAllEndpoints();

    expect(endpoints).not.toBeNull();
  });

  it('Should return version of the selected endpoint', () => {
    const version = QorusAuthenticator.getEndpointVersion();

    expect(version).toEqual('latest');
  });

  it('Should return api paths for the selected endpoint', () => {
    expect(QorusAuthenticator.getApiPaths()).toMatchSnapshot();
  });

  it('Should set a new version for the endpoint', async () => {
    expect(await QorusAuthenticator.setEndpointVersion(5)).toMatchSnapshot();
  });

  it('Should revalidate the user auth token for the selected endpoint', async () => {
    await QorusAuthenticator.setEndpointVersion('latest');
    await QorusAuthenticator.renewSelectedEndpointToken({ user: process.env.TESTUSER!, pass: process.env.TESTPASS! });
    const token = QorusAuthenticator.getAuthToken();

    expect(typeof token).toEqual('string');
  });

  it('Should return current user token if the user is authenticated', () => {
    const token = QorusAuthenticator.getAuthToken();

    expect(typeof token).toEqual('string');
  });

  it('Should return the current endpoint', () => {
    const config = QorusAuthenticator.getSelectedEndpoint();

    expect(config!.endpointId).toMatchSnapshot();
  });

  it('Should return all the endpoints', () => {
    const endpoints = QorusAuthenticator.getAllEndpoints();

    expect(endpoints.length).toMatchSnapshot();
  });

  it('Should change the selected endpoint url and logout the user', async () => {
    const url = await QorusAuthenticator.setEndpointUrl('https://testme.com');

    expect(url).toMatchSnapshot();
    expect(QorusAuthenticator.getAuthToken()).toBeUndefined();
  });

  it('Should select the endpoint by the provided id', async () => {
    if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT, endpointId: 'test' });

    expect(QorusAuthenticator.selectEndpoint('test')).toMatchSnapshot();
  });

  // it.only('Should create a new endpoint with port accessible', async () => {
  //   let endpoint: Endpoint | undefined;
  //   if (process.env.ENDPOINT) {
  //     endpoint = QorusAuthenticator.addEndpoint({
  //       url: 'https://hq.qoretechnologies.com:31011',
  //       id: 'rippyPort',
  //     });
  //   }
  //   await QorusAuthenticator.login({ user: process.env.TESTUSER!, pass: process.env.TESTPASS! });
  //   expect(typeof endpoint?.authToken).toEqual('string');
  // });
});

describe('QorusLogin Utility Error Tests', () => {
  it('Should throw an Authentication Error when user tries to authenticate with wrong credentials', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT, endpointId: 'rippy2' });
      await QorusAuthenticator.login({ user: 'bob', pass: 'pass' });
    } catch (error) {
      err = error;
    }
    console.log(err);
    expect(err instanceof ErrorQorusRequest).toEqual(true);
  });

  it('Should throw an error if the user does not provide username and password for authentication.', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: process.env.ENDPOINT, endpointId: 'rippy2' });
      await QorusAuthenticator.login();
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal Error if the id and url are not valid to initialize an endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT) QorusAuthenticator.addEndpoint({ url: '', endpointId: '' });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Authentication error if the username and pass is not valid while initializing endpoint', async () => {
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: process.env.ENDPOINT,
          endpointId: 'rippy2',
        });
      await QorusAuthenticator.login({
        user: '',
        pass: '',
      });
    } catch (error) {
      console.error(error);
    }
    expect(loggerMock).toHaveBeenCalled();
  });

  it('Should throw an Internal error if the no-auth status cannot be checked', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: 'https://sandbox.qoretechnologies.',
          endpointId: 'rippy2',
        });
      await QorusAuthenticator.login({
        user: 'some',
        pass: 'some',
      });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if url is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: '',
          endpointId: 'rippy2',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if id is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: 'https://sandbox.qoretechnologies.com',
          endpointId: '',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });

  it('Should throw an Internal error if id and url is not valid for initializing endpoint', async () => {
    let err;
    try {
      if (process.env.ENDPOINT)
        QorusAuthenticator.addEndpoint({
          url: '',
          endpointId: '',
        });
    } catch (error) {
      err = error;
    }
    expect(err instanceof ErrorInternal).toEqual(true);
  });
});
