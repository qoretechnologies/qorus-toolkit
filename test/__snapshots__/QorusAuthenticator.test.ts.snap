// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QorusLogin Utility Class Tests Should change the selected endpoint url and logout the user 1`] = `"https://testme.com"`;

exports[`QorusLogin Utility Class Tests Should initialize the endpoint and assign it to the selected endpoint 1`] = `
{
  "endpointId": "rippy",
  "url": "https://sandbox.qoretechnologies.com",
  "version": "latest",
}
`;

exports[`QorusLogin Utility Class Tests Should return all the endpoints 1`] = `1`;

exports[`QorusLogin Utility Class Tests Should return api paths for the selected endpoint 1`] = `
{
  "authenticator": {
    "login": "/api/latest/public/login",
    "logout": "/api/latest/logout",
    "validateNoAuth": "/api/latest/public/info",
    "validateToken": "/api/latest/system?action=validateWsToken",
  },
  "dataProviders": {
    "browse": "/api/latest/dataprovider/browse",
  },
  "jobs": {
    "browse": "/api/latest/jobs",
  },
  "workflows": {
    "browse": "/api/latest/workflows",
  },
}
`;

exports[`QorusLogin Utility Class Tests Should return the current endpoint 1`] = `"rippy"`;

exports[`QorusLogin Utility Class Tests Should select the endpoint by the provided id 1`] = `Promise {}`;

exports[`QorusLogin Utility Class Tests Should set a new version for the endpoint 1`] = `5`;
